internal/relationships/validation.go

```diff
@@ -3,22 +3,86 @@ package relationships
 import (
 	"context"
 
+	"github.com/samber/lo"
+
 	"github.com/authzed/spicedb/internal/namespace"
 	"github.com/authzed/spicedb/pkg/caveats"
 	"github.com/authzed/spicedb/pkg/datastore"
 	"github.com/authzed/spicedb/pkg/genutil/mapz"
 	ns "github.com/authzed/spicedb/pkg/namespace"
 	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+	"github.com/authzed/spicedb/pkg/spiceerrors"
 	"github.com/authzed/spicedb/pkg/tuple"
 )
 
-// ValidateRelationships performs validation on the given relationships, ensuring that
+// ValidateRelationshipUpdates performs validation on the given relationship updates, ensuring that
 // they can be applied against the datastore.
-func ValidateRelationships(
+func ValidateRelationshipUpdates(
+	ctx context.Context,
+	reader datastore.Reader,
+	updates []*core.RelationTupleUpdate,
+) error {
+	rels := lo.Map(updates, func(item *core.RelationTupleUpdate, _ int) *core.RelationTuple {
+		return item.Tuple
+	})
+
+	// Load namespaces and caveats.
+	referencedNamespaceMap, referencedCaveatMap, err := loadNamespacesAndCaveats(ctx, rels, reader)
+	if err != nil {
+		return err
+	}
+
+	// Validate each updates's types.
+	for _, update := range updates {
+		option := ValidateRelationshipForWriting
+		if update.Operation == core.RelationTupleUpdate_DELETE {
+			option = ValidateRelationshipForDeletion
+		}
+
+		if err := ValidateOneRelationship(
+			referencedNamespaceMap,
+			referencedCaveatMap,
+			update.Tuple,
+			option,
+		); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ValidateRelationshipsForCreateOrTouch performs validation on the given relationships to be written, ensuring that
+// they can be applied against the datastore.
+//
+// NOTE: This method *cannot* be used for relationships that will be deleted.
+func ValidateRelationshipsForCreateOrTouch(
```

Comment:
Either call this `...ForWrite` or rename the constant to be `...ForCreateOrTouch`. I lean towards the latter.

Reply:
Done